components/camel-grpc/src/main/java/org/apache/camel/component/grpc/GrpcConsumer.java

```diff
@@ -0,0 +1,142 @@
+    }
+    
+    public GrpcConfiguration getConfiguration() {
+        return configuration;
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        super.doStart();
+        if (server == null) {
+            LOG.info("Starting the gRPC server");
+            initializeServer();
+            server.start();
+            LOG.info("gRPC server started and listening on port: " + server.getPort());
+        }
+    }
+
+    @Override
+    protected void doStop() throws Exception {
+        if (server != null) {
+            LOG.trace("Terminating gRPC server");
+            server.shutdown().shutdownNow();
+            server = null;
+        }
+        super.doStop();
+    }
+
+    protected void initializeServer() {
+        NettyServerBuilder serverBuilder = null;
+        BindableService bindableService = null;
+        ProxyFactory serviceProxy = new ProxyFactory();
+        ServerInterceptor headerInterceptor = new GrpcHeaderInterceptor();
+        MethodHandler methodHandler = new GrpcMethodHandler(endpoint, this);
+        
+        serviceProxy.setSuperclass(GrpcUtils.constructGrpcImplBaseClass(configuration.getServicePackage(), configuration.getServiceName()));
+        try {
+            bindableService = (BindableService)serviceProxy.create(new Class<?>[0], new Object[0], methodHandler);
+        } catch (NoSuchMethodException | IllegalArgumentException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            throw new IllegalArgumentException("Unable to create bindable proxy service for " + configuration.getService());
+        }
+        
+        if (!ObjectHelper.isEmpty(configuration.getHost()) && !ObjectHelper.isEmpty(configuration.getPort())) {
+            LOG.info("Building gRPC server on " + configuration.getHost() + ":" + configuration.getPort());
+            serverBuilder = NettyServerBuilder.forAddress(new InetSocketAddress(configuration.getHost(), configuration.getPort()));
+        } else if (ObjectHelper.isEmpty(configuration.getHost()) && !ObjectHelper.isEmpty(configuration.getPort())) {
+            LOG.info("Building gRPC server on <any address>" + ":" + configuration.getPort());
+            serverBuilder = NettyServerBuilder.forPort(configuration.getPort());
+        } else {
+            throw new IllegalArgumentException("No server start properties (host, port) specified");
+        }
+        
+        server = serverBuilder.addService(ServerInterceptors.intercept(bindableService, headerInterceptor)).build();
+    }
+    
+    public boolean process(Exchange exchange, AsyncCallback callback) {
+        exchange.getIn().setHeader(GrpcConstants.GRPC_EVENT_TYPE_HEADER, GrpcConstants.GRPC_EVENT_TYPE_ON_NEXT);
+        return doSend(exchange, callback);
+    }
+    
+    public void onCompleted(Exchange exchange) {
+        exchange.getIn().setHeader(GrpcConstants.GRPC_EVENT_TYPE_HEADER, GrpcConstants.GRPC_EVENT_TYPE_ON_COMPLETED);
```

Comment:
Probably it's better adding some flags (defaulting to false) to avoid forwarding empty exchanges with these information unless they are required by the users. Something like: https://github.com/apache/camel/blob/master/components/camel-reactive-streams/src/main/java/org/apache/camel/component/reactive/streams/ReactiveStreamsEndpoint.java#L50-L54